openstef/model/metamodels/grouped_regressor.py

```diff
@@ -0,0 +1,141 @@
+    Moreover fit and predict methods can be performed in parallel for each group key thanks to joblib.
+    Example:
+
+    data = | index | group | x0 | x1 | x3 | y |
+           |   0   |   1   | .. | .. | .. | . |
+           |   1   |   2   | .. | .. | .. | . |
+           |   2   |   1   | .. | .. | .. | . |
+           |   3   |   2   | .. | .. | .. | . |
+
+           [              X              ][ Y ]
+
+    The GroupedRegressor on the data with the group_columns='group' fits 2 models :
+        the model 1 with the row 0 and 2, columns x0, x1 and x3 as the features and column y as the target.
+        the model 2 with the row 1 and 3, columns x0, x1 and x3 as the features and column y as the target.
+
+    Parameters
+    ----------
+    base_estimator: RegressorMixin
+        Regressor .
+
+    group_columns : str, int, list
+        Name(s) of the column(s) used as the key for groupby operation.
+
+    n_jobs : int default=1
+        The maximum number of concurrently running jobs,
+         such as the number of Python worker processes when backend=”multiprocessing”
+         or the size of the thread-pool when backend=”threading
+
+     Attributes
+    ----------
+        feature_names_: list(str)
+            All input feature (without group_columns).
+
+        estimators_: dict(str, RegressorMixin)
+            Dictionnary that stocks fitted estimators for each group.
+            The keys are the keys of grouping and the values are the regressors fitted on the grouped data.
+    """
+
+    def __init__(self, base_estimator, group_columns, n_jobs=1):
+        self.base_estimator = base_estimator
+        if type(group_columns) in [int, str]:
+            self.group_columns = [group_columns]
+        else:
+            self.group_columns = group_columns
+        self.n_jobs = 1
+
+    def _get_tags(self):
+        return self.base_estimator._get_tags()
+
+    def _check_group_columns(self, df):
+        if type(self.group_columns) is not list:
+            raise ValueError(
+                "The group columns parameter should be a list, it gets a {}".format(
+                    type(self.group_columns)
+                )
+            )
+        for c in self.group_columns:
+            if c not in list(df.columns):
+                raise ValueError("The group column {} is missing!".format(c))
+
+    def _partial_fit(self, group, df_group):
+        estimator = clone(self.base_estimator)
+        X = df_group.loc[:, self.feature_names_]
+        y = df_group.loc[:, "__target__"]
+        return (group, estimator.fit(X, y))
+
+    def _partial_predict(self, group, df_group):
+        return self.estimators_[group].predict(df_group)
+
+    @classmethod
+    def grouped_compute(cls, df, group_columns, func, n_jobs=1):
+        index_name = df.index.name or "index"
+        df_reset = df.reset_index()
+
+        df_res = pd.DataFrame(index=df_reset.index)
+
+        gb = df_reset.groupby(group_columns)
+
+        if n_jobs > 1:
```

Comment:
Nice to make this parallel, do you think it is a good idea to also implement this in de xgb_quantile model?

Reply:
Yes we think that the xgb_quantile could be implemented with the grouped_regressor. However if the current implementation is working properly we suggest that it stays like that until you have time to test the switch. (If we have time we can propose to you an implementation but it is not in our priorities)

Reply:
Great that we could later switch to use grouped_regressor too